<a href="https://colab.research.google.com/github/anurag1210/ML_ZoomCamp_Homework_2024/blob/main/Homework_6_Decision_Trees%26EnsambleLearnings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Importing the necessary packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
#Loading the Dataset
!wget https://github.com/alexeygrigorev/datasets/raw/refs/heads/master/jamb_exam_results.csv

--2024-11-06 20:36:08--  https://github.com/alexeygrigorev/datasets/raw/refs/heads/master/jamb_exam_results.csv
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/alexeygrigorev/datasets/refs/heads/master/jamb_exam_results.csv [following]
--2024-11-06 20:36:08--  https://raw.githubusercontent.com/alexeygrigorev/datasets/refs/heads/master/jamb_exam_results.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 391501 (382K) [text/plain]
Saving to: ‘jamb_exam_results.csv.1’

jamb_exam_results.c 100%[===================>] 382.33K  --.-KB/s    in 0.05s   

2024-11-06 20:36:08 (6.90 MB/s) - ‘jamb_exam_resu

In [ ]:
#Reading the Dataset
df = pd.read_csv('jamb_exam_results.csv')

In [ ]:
df.head()

JAMB_Score  Study_Hours_Per_Week  Attendance_Rate  Teacher_Quality  \
0         192                    22               78                4   
1         207                    14               88                4   
2         182                    29               87                2   
3         210                    29               99                2   
4         199                    12               98                3   

   Distance_To_School School_Type School_Location Extra_Tutorials  \
0                12.4      Public           Urban             Yes   
1                 2.7      Public           Rural              No   
2                 9.6      Public           Rural             Yes   
3                 2.6      Public           Urban              No   
4                 8.8      Public           Urban              No   

  Access_To_Learning_Materials Parent_Involvement IT_Knowledge  Student_ID  \
0                          Yes               High       Medium           1   
1                          Yes               High         High           2   
2                          Yes               High       Medium           3   
3                          Yes             Medium         High           4   
4                          Yes             Medium       Medium           5   

   Age  Gender Socioeconomic_Status Parent_Education_Level  \
0   17    Male                  Low               Tertiary   
1   15    Male                 High                    NaN   
2   20  Female                 High               Tertiary   
3   22  Female               Medium               Tertiary   
4   22  Female               Medium               Tertiary   

   Assignments_Completed  
0                      2  
1                      1  
2                      2  
3                      1  
4                      1

In [ ]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [ ]:
df.head()

jamb_score  study_hours_per_week  attendance_rate  teacher_quality  \
0         192                    22               78                4   
1         207                    14               88                4   
2         182                    29               87                2   
3         210                    29               99                2   
4         199                    12               98                3   

   distance_to_school school_type school_location extra_tutorials  \
0                12.4      Public           Urban             Yes   
1                 2.7      Public           Rural              No   
2                 9.6      Public           Rural             Yes   
3                 2.6      Public           Urban              No   
4                 8.8      Public           Urban              No   

  access_to_learning_materials parent_involvement it_knowledge  student_id  \
0                          Yes               High       Medium           1   
1                          Yes               High         High           2   
2                          Yes               High       Medium           3   
3                          Yes             Medium         High           4   
4                          Yes             Medium       Medium           5   

   age  gender socioeconomic_status parent_education_level  \
0   17    Male                  Low               Tertiary   
1   15    Male                 High                    NaN   
2   20  Female                 High               Tertiary   
3   22  Female               Medium               Tertiary   
4   22  Female               Medium               Tertiary   

   assignments_completed  
0                      2  
1                      1  
2                      2  
3                      1  
4                      1

#Preparation:

#Remove the student_id column.
#Fill missing values with zeros.
#Do train/validation/test split with 60%/20%/20% distribution.
#Use the train_test_split function and set the random_state parameter to 1.
#Use DictVectorizer(sparse=True) to turn the dataframes into matrices.

In [ ]:
#Removing the  student_id column
# axis =0 refers to row and axis = 1 refers to column
df = df.drop('student_id', axis=1)

In [ ]:
df.head()

jamb_score  study_hours_per_week  attendance_rate  teacher_quality  \
0         192                    22               78                4   
1         207                    14               88                4   
2         182                    29               87                2   
3         210                    29               99                2   
4         199                    12               98                3   

   distance_to_school school_type school_location extra_tutorials  \
0                12.4      Public           Urban             Yes   
1                 2.7      Public           Rural              No   
2                 9.6      Public           Rural             Yes   
3                 2.6      Public           Urban              No   
4                 8.8      Public           Urban              No   

  access_to_learning_materials parent_involvement it_knowledge  age  gender  \
0                          Yes               High       Medium   17    Male   
1                          Yes               High         High   15    Male   
2                          Yes               High       Medium   20  Female   
3                          Yes             Medium         High   22  Female   
4                          Yes             Medium       Medium   22  Female   

  socioeconomic_status parent_education_level  assignments_completed  
0                  Low               Tertiary                      2  
1                 High                    NaN                      1  
2                 High               Tertiary                      2  
3               Medium               Tertiary                      1  
4               Medium               Tertiary                      1

In [ ]:
#Checking missing values
df.isnull().sum()

jamb_score                        0
study_hours_per_week              0
attendance_rate                   0
teacher_quality                   0
distance_to_school                0
school_type                       0
school_location                   0
extra_tutorials                   0
access_to_learning_materials      0
parent_involvement                0
it_knowledge                      0
age                               0
gender                            0
socioeconomic_status              0
parent_education_level          891
assignments_completed             0
dtype: int64

From the above we see the parent_education_level column has 891 records as missing values

In [ ]:
#Filling the parent_education_level null values with zeroes
df = df.fillna(0)

In [ ]:
df.isnull().sum()

jamb_score                      0
study_hours_per_week            0
attendance_rate                 0
teacher_quality                 0
distance_to_school              0
school_type                     0
school_location                 0
extra_tutorials                 0
access_to_learning_materials    0
parent_involvement              0
it_knowledge                    0
age                             0
gender                          0
socioeconomic_status            0
parent_education_level          0
assignments_completed           0
dtype: int64

In [ ]:
df.head(10)

jamb_score  study_hours_per_week  attendance_rate  teacher_quality  \
0         192                    22               78                4   
1         207                    14               88                4   
2         182                    29               87                2   
3         210                    29               99                2   
4         199                    12               98                3   
5         202                    25               85                2   
6         251                    35               85                4   
7         129                    27               75                3   
8         220                    23               85                3   
9         157                    15               79                3   

   distance_to_school school_type school_location extra_tutorials  \
0                12.4      Public           Urban             Yes   
1                 2.7      Public           Rural              No   
2                 9.6      Public           Rural             Yes   
3                 2.6      Public           Urban              No   
4                 8.8      Public           Urban              No   
5                13.6      Public           Urban             Yes   
6                 2.6      Public           Urban              No   
7                 9.4      Public           Urban              No   
8                 4.6      Public           Rural              No   
9                15.6      Public           Rural              No   

  access_to_learning_materials parent_involvement it_knowledge  age  gender  \
0                          Yes               High       Medium   17    Male   
1                          Yes               High         High   15    Male   
2                          Yes               High       Medium   20  Female   
3                          Yes             Medium         High   22  Female   
4                          Yes             Medium       Medium   22  Female   
5                           No             Medium          Low   15    Male   
6                          Yes                Low       Medium   16  Female   
7                          Yes                Low       Medium   19  Female   
8                           No                Low       Medium   19  Female   
9                          Yes                Low          Low   20  Female   

  socioeconomic_status parent_education_level  assignments_completed  
0                  Low               Tertiary                      2  
1                 High                      0                      1  
2                 High               Tertiary                      2  
3               Medium               Tertiary                      1  
4               Medium               Tertiary                      1  
5                  Low               Tertiary                      1  
6               Medium                Primary                      4  
7                  Low               Tertiary                      3  
8               Medium               Tertiary                      1  
9               Medium                      0                      1

In [ ]:
#splitting the train/validation/test split  of 60/20/20 distribution
from sklearn.model_selection import train_test_split

# First split: 60% training and 40% remaining (for validation and test)
train, temp = train_test_split(df, train_size=0.6, random_state=1)

# Second split: 20% validation and 20% test from the remaining 40%
val, test = train_test_split(temp, test_size=0.5, random_state=1)

# Check the shapes to confirm the split
print("Train set:", train.shape)
print("Validation set:", val.shape)
print("Test set:", test.shape)

#Train set: (3000, 16)
#Validation set: (1000, 16)
#Test set: (1000, 16)

Train set: (3000, 16)
Validation set: (1000, 16)
Test set: (1000, 16)


Q1: Let's train a decision tree regressor to predict the jamb_score variable.

Train a model with max_depth=1.
Which feature is used for splitting the data?

study_hours_per_week
attendance_rate
teacher_quality
distance_to_school


In [ ]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.tree import DecisionTreeRegressor

# Remove target variable from DataFrames
X_train_df = train.drop(columns=['jamb_score'])
X_val_df = val.drop(columns=['jamb_score'])
X_test_df = test.drop(columns=['jamb_score'])

# Convert each DataFrame to a list of dictionaries
train_dict = X_train_df.to_dict(orient='records')
val_dict = X_val_df.to_dict(orient='records')
test_dict = X_test_df.to_dict(orient='records')

# Initialize DictVectorizer
vec = DictVectorizer(sparse=True)

# Fit and transform the train set, then transform the validation and test sets
X_train = vec.fit_transform(train_dict)
X_val = vec.transform(val_dict)
X_test = vec.transform(test_dict)

# Extract y_train and y_val separately
y_train = train['jamb_score']
y_val = val['jamb_score']

# Train the Decision Tree with max_depth=1
model = DecisionTreeRegressor(max_depth=1, random_state=1)
model.fit(X_train, y_train)

# Retrieve the feature used for the split
feature_index = model.tree_.feature[0]
feature_name = vec.feature_names_[feature_index]

print("Feature used for splitting:", feature_name)

Feature used for splitting: study_hours_per_week


Q2 Train a random forest model with these parameters:

n_estimators=10
random_state=1
n_jobs=-1 (optional - to make training faster)
What's the RMSE of this model on validation?

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

# Train the Random Forest model
rf_model = RandomForestRegressor(n_estimators=10, random_state=1, n_jobs=-1)
rf_model.fit(X_train, y_train)

# Predict on the validation set
y_pred = rf_model.predict(X_val)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_val, y_pred))

print("RMSE on the validation data:", rmse)

RMSE on the validation data: 43.157758977963624


Q3. Now let's experiment with the n_estimators parameter

Try different values of this parameter from 10 to 200 with step 10.
Set random_state to 1.
Evaluate the model on the validation dataset.
After which value of n_estimators does RMSE stop improving? Consider 3 decimal places for calculating the answer.


In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

# Initialize variables
n_estimators_values = range(10, 201, 10)  # Values from 10 to 200 with step 10
rmse_values = []  # To store the RMSE for each n_estimators

# Loop through each value of n_estimators
for n_estimators in n_estimators_values:
    rf_model = RandomForestRegressor(n_estimators=n_estimators, random_state=1, n_jobs=-1)
    rf_model.fit(X_train, y_train)

    # Predict on the validation set
    y_pred = rf_model.predict(X_val)

    # Calculate RMSE
    rmse = np.sqrt(mean_squared_error(y_val, y_pred))
    rmse_values.append(rmse)

    print(f"n_estimators: {n_estimators}, RMSE: {rmse}")

# Find the point where RMSE stops improving
for i in range(1, len(rmse_values)):
    if round(rmse_values[i], 3) == round(rmse_values[i-1], 3):
        print(f"RMSE stops improving after n_estimators = {n_estimators_values[i-1]}")
        break

n_estimators: 10, RMSE: 43.157758977963624
n_estimators: 20, RMSE: 41.79043981582391
n_estimators: 30, RMSE: 41.555818472133225
n_estimators: 40, RMSE: 41.075631652173044
n_estimators: 50, RMSE: 40.9571573818301
n_estimators: 60, RMSE: 40.77368529456223
n_estimators: 70, RMSE: 40.587805985220214
n_estimators: 80, RMSE: 40.5027042403498
n_estimators: 90, RMSE: 40.43492224596255
n_estimators: 100, RMSE: 40.36491034549687
n_estimators: 110, RMSE: 40.347525479439874
n_estimators: 120, RMSE: 40.30191844844362
n_estimators: 130, RMSE: 40.285789466741
n_estimators: 140, RMSE: 40.26346078629849
n_estimators: 150, RMSE: 40.25426440073703
n_estimators: 160, RMSE: 40.1996656828838
n_estimators: 170, RMSE: 40.187325737485885
n_estimators: 180, RMSE: 40.13596272032919
n_estimators: 190, RMSE: 40.15216599857013
n_estimators: 200, RMSE: 40.138465594427


Question 4
Let's select the best max_depth:

Try different values of max_depth: [10, 15, 20, 25]
For each of these values,
try different values of n_estimators from 10 till 200 (with step 10)
calculate the mean RMSE
Fix the random seed: random_state=1
What's the best max_depth, using the mean RMSE?


Question 4
Let's select the best max_depth:

Try different values of max_depth: [10, 15, 20, 25]
For each of these values,
try different values of n_estimators from 10 till 200 (with step 10)
calculate the mean RMSE
Fix the random seed: random_state=1
What's the best max_depth, using the mean RMSE?

In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

# Initialize variables
max_depth_values = [10, 15, 20, 25]
n_estimators_values = range(10, 201, 10)  # Values from 10 to 200 with step 10
mean_rmse_per_depth = {}  # To store the mean RMSE for each max_depth

# Loop through each max_depth value
for max_depth in max_depth_values:
    rmse_values = []  # To store the RMSE for each n_estimators
    for n_estimators in n_estimators_values:
        rf_model = RandomForestRegressor(n_estimators=n_estimators,
                                         max_depth=max_depth,
                                         random_state=1,
                                         n_jobs=-1)
        rf_model.fit(X_train, y_train)

        # Predict on the validation set
        y_pred = rf_model.predict(X_val)

        # Calculate RMSE
        rmse = np.sqrt(mean_squared_error(y_val, y_pred))
        rmse_values.append(rmse)

    # Calculate the mean RMSE for this max_depth
    mean_rmse = np.mean(rmse_values)
    mean_rmse_per_depth[max_depth] = mean_rmse
    print(f"max_depth: {max_depth}, Mean RMSE: {mean_rmse}")

# Find the best max_depth (lowest mean RMSE)
best_max_depth = min(mean_rmse_per_depth, key=mean_rmse_per_depth.get)
print(f"The best max_depth is: {best_max_depth} with a mean RMSE of {mean_rmse_per_depth[best_max_depth]}")

max_depth: 10, Mean RMSE: 40.13800364122155
max_depth: 15, Mean RMSE: 40.64396557212339
max_depth: 20, Mean RMSE: 40.60981968462706


Question 5
We can extract feature importance information from tree-based models.

At each step of the decision tree learning algorithm, it finds the best split. When doing it, we can calculate "gain" - the reduction in impurity before and after the split. This gain is quite useful in understanding what are the important features for tree-based models.

In Scikit-Learn, tree-based models contain this information in the feature_importances_ field.

For this homework question, we'll find the most important feature:

Train the model with these parameters:
n_estimators=10,
max_depth=20,
random_state=1,
n_jobs=-1 (optional)
Get the feature importance information from this model
What's the most important feature (among these 4)?

study_hours_per_week
attendance_rate
distance_to_school
teacher_quality

In [59]:
from sklearn.ensemble import RandomForestRegressor

# Initialize the model with the given parameters
model = RandomForestRegressor(n_estimators=10,
                              max_depth=20,
                              random_state=1,
                              n_jobs=-1)

# Train the model
model.fit(X_train, y_train)

# Get the feature importance values
feature_importances = model.feature_importances_

# Use vec.feature_names_ to get the feature names from DictVectorizer
features = vec.feature_names_

# Create a dictionary of feature names and their corresponding importance
feature_importance_dict = dict(zip(features, feature_importances))

# Print the feature importance
print("Feature importances:", feature_importance_dict)

# Find the most important feature
most_important_feature = max(feature_importance_dict, key=feature_importance_dict.get)
print("Most important feature:", most_important_feature)

Feature importances: {'access_to_learning_materials=No': 0.009913440694149447, 'access_to_learning_materials=Yes': 0.008560644856282888, 'age': 0.06439509739395811, 'assignments_completed': 0.03442850609469451, 'attendance_rate': 0.15213485818850261, 'distance_to_school': 0.13576097373953083, 'extra_tutorials=No': 0.011600937761865035, 'extra_tutorials=Yes': 0.010880314324883043, 'gender=Female': 0.009655439345357855, 'gender=Male': 0.010536816837096813, 'it_knowledge=High': 0.01627493805822261, 'it_knowledge=Low': 0.013348787522459549, 'it_knowledge=Medium': 0.011727154148302247, 'parent_education_level': 0.0, 'parent_education_level=Primary': 0.012055837569125174, 'parent_education_level=Secondary': 0.014631753124537166, 'parent_education_level=Tertiary': 0.014306214122188193, 'parent_involvement=High': 0.025248894866975284, 'parent_involvement=Low': 0.016064340357081305, 'parent_involvement=Medium': 0.011179765424027654, 'school_location=Rural': 0.010070153938938288, 'school_locatio

 Question 6
Now let's train an XGBoost model! For this question, we'll tune the eta parameter:

Install XGBoost
Create DMatrix for train and validation
Create a watchlist
Train a model with these parameters for 100 rounds:
xgb_params = {
    'eta': 0.3,
    'max_depth': 6,
    'min_child_weight': 1,
    
    'objective': 'reg:squarederror',
    'nthread': 8,
    
    'seed': 1,
    'verbosity': 1,
}
Now change eta from 0.3 to 0.1.

Which eta leads to the best RMSE score on the validation dataset?

0.3
0.1
Both give equal value

In [60]:
!pip install xgboost

In [61]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error

# Convert your data into DMatrix format for XGBoost
dtrain = xgb.DMatrix(X_train, label=y_train)
dval = xgb.DMatrix(X_val, label=y_val)

# Create a watchlist to evaluate the model on the validation set during training
watchlist = [(dtrain, 'train'), (dval, 'validation')]

In [62]:
# Define parameters with eta = 0.3
xgb_params_0_3 = {
    'eta': 0.3,
    'max_depth': 6,
    'min_child_weight': 1,
    'objective': 'reg:squarederror',
    'nthread': 8,
    'seed': 1,
    'verbosity': 1,
}

# Train the model
model_0_3 = xgb.train(xgb_params_0_3, dtrain, num_boost_round=100, evals=watchlist, early_stopping_rounds=10)

[0]	train-rmse:42.84835	validation-rmse:44.52338
[1]	train-rmse:39.96423	validation-rmse:42.83406
[2]	train-rmse:37.91231	validation-rmse:41.62607
[3]	train-rmse:36.51126	validation-rmse:41.25491
[4]	train-rmse:35.52212	validation-rmse:40.84075
[5]	train-rmse:34.77126	validation-rmse:40.71677
[6]	train-rmse:34.03898	validation-rmse:40.72669
[7]	train-rmse:33.62820	validation-rmse:40.68822
[8]	train-rmse:32.94729	validation-rmse:40.81273
[9]	train-rmse:32.27703	validation-rmse:40.84939
[10]	train-rmse:31.73818	validation-rmse:40.83759
[11]	train-rmse:31.31360	validation-rmse:40.80575
[12]	train-rmse:30.72949	validation-rmse:40.84238
[13]	train-rmse:30.11486	validation-rmse:40.96020
[14]	train-rmse:29.43538	validation-rmse:40.98775
[15]	train-rmse:29.23018	validation-rmse:41.04798
[16]	train-rmse:28.64113	validation-rmse:41.08375


In [63]:
# Define parameters with eta = 0.1
xgb_params_0_1 = {
    'eta': 0.1,
    'max_depth': 6,
    'min_child_weight': 1,
    'objective': 'reg:squarederror',
    'nthread': 8,
    'seed': 1,
    'verbosity': 1,
}

# Train the model
model_0_1 = xgb.train(xgb_params_0_1, dtrain, num_boost_round=100, evals=watchlist, early_stopping_rounds=10)


[0]	train-rmse:45.64414	validation-rmse:46.63724
[1]	train-rmse:44.26862	validation-rmse:45.58724
[2]	train-rmse:43.08569	validation-rmse:44.76209
[3]	train-rmse:42.05227	validation-rmse:44.02498
[4]	train-rmse:41.10533	validation-rmse:43.40640
[5]	train-rmse:40.28309	validation-rmse:42.92195
[6]	train-rmse:39.54133	validation-rmse:42.49211
[7]	train-rmse:38.87686	validation-rmse:42.15780
[8]	train-rmse:38.27674	validation-rmse:41.84104
[9]	train-rmse:37.74058	validation-rmse:41.58026
[10]	train-rmse:37.26338	validation-rmse:41.35829
[11]	train-rmse:36.82810	validation-rmse:41.19143
[12]	train-rmse:36.41091	validation-rmse:41.02571
[13]	train-rmse:36.01019	validation-rmse:40.90308
[14]	train-rmse:35.67454	validation-rmse:40.79701
[15]	train-rmse:35.33492	validation-rmse:40.66274
[16]	train-rmse:35.01425	validation-rmse:40.60840
[17]	train-rmse:34.72687	validation-rmse:40.55942
[18]	train-rmse:34.40588	validation-rmse:40.46321
[19]	train-rmse:34.16207	validation-rmse:40.42760
[20]	train

In [64]:
# Get the best RMSE scores from each model
rmse_0_3 = model_0_3.best_score
rmse_0_1 = model_0_1.best_score

print(f"Best RMSE with eta=0.3: {rmse_0_3}")
print(f"Best RMSE with eta=0.1: {rmse_0_1}")


Best RMSE with eta=0.3: 40.68821969954672
Best RMSE with eta=0.1: 40.166449496198915
